broadcaster/broadcaster.go

```diff
@@ -0,0 +1,162 @@
+package broadcaster
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"github.com/spacemeshos/poet/broadcaster/pb"
+	"github.com/spacemeshos/smutil/log"
+	"google.golang.org/grpc"
+	"sync"
+	"time"
+)
+
+const (
+	connectionTimeout = 10 * time.Second
+	broadcastTimeout  = 30 * time.Second
+)
+
+type Broadcaster struct {
+	clients     []pb.SpacemeshServiceClient
+	connections []*grpc.ClientConn
+}
+
+func (b *Broadcaster) BroadcastProof(msg []byte, roundId string, members [][]byte) error {
+	if b.clients == nil {
+		log.Info("Broadcast is disabled, not broadcasting round %v proof", roundId)
+		return nil
+	}
+	pbMsg := &pb.BinaryMessage{Data: msg}
+	ctx, cancel := context.WithTimeout(context.Background(), broadcastTimeout)
+	defer cancel()
+
+	responses := make([]*pb.SimpleMessage, len(b.clients))
+	errs := make([]error, len(b.clients))
+
+	start := time.Now()
+	var wg sync.WaitGroup
+	wg.Add(len(b.clients))
+	for i, client := range b.clients {
+		i := i
+		client := client
+		go func() {
+			defer wg.Done()
+			responses[i], errs[i] = client.BroadcastPoet(ctx, pbMsg)
+		}()
+	}
+	wg.Wait()
+	elapsed := time.Since(start)
+
+	// Parse and concatenate potential errors.
+	var numErrors int
+	var retErr error
+	for i := range b.clients {
+		err := errs[i]
+		res := responses[i]
+		if err != nil {
+			err = fmt.Errorf("failed to broadcast via gateway node at \"%v\" after %v: %v", b.connections[i].Target(), elapsed, err)
+		} else if res.Value != "ok" {
```

Comment:
Minor, style, ignore if you disagree:
```suggestion
		} else if res := responses[i]; res.Value != "ok" {
```
Instead of setting `res` in advance, even if there's an `err`.

Reply:
IMO this optimization produces a slightly less readable code, so I'd rather leave it.